text/java/0000-apache-tomcat-launch-config.md

```diff
@@ -0,0 +1,60 @@
+# Apache Tomcat launch configuration
+
+## Summary
+
+With the current implementation of the Apache Tomcat buildpack, its quite difficult to configure values at launch time that affect values in the `web.xml`, `server.xml` & `context.xml`. We 
+want to provide a way to make some features easily configurable and controlled at launch time rather than build time.
+
+## Motivation
+
+The Apache Tomcat buildpack provides a method of adding a configuration overlay using the build time `BP_TOMCAT_EXT_CONF_*` variables, this takes a pre-built archive
+and extracts this during build time. This method can be cumbersome if there are values that need to be changed on a per-deployment basis.  Apache Tomcat already supports
+the use of [system property replacements](https://tomcat.apache.org/tomcat-9.0-doc/config/systemprops.html#Property_replacements) within its configuration files.  The aim is to make
+the configuring of the system properties at launch time lot easier.
+ 
+## Detailed Explanation
+
+Given an example `context.xml`, which can be included in either an external archive, or inside the applications `META-INF` directory:
+
+```
+<Context>
+    <Resources allowLinking="true"/>
+
+    <Resource name="jdbc/Datasource" auth="Container"
+              type="javax.sql.DataSource" driverClassName="org.postgresql.Driver"
+              url="${postgres.url}" username="${postgres.username}"
+              password="${postgres.password}" maxTotal="25"
+              maxIdle="10" validationQuery="select 1"/>
+
+</Context>
+```
+
+The resource configuration is static apart from the `url`, `username` and `password` parameters, but we do not want these to be written into the image at build time.
+
+## Rationale and Alternatives
+
+It is possible to use `BPE_*` parameters with the [environment variables buildpack](https://github.com/paketo-buildpacks/environment-variables), these are written into the image at buildtime.
+
+These values could also be configured by updating `JAVA_TOOL_OPTIONS` variable, but this can be quite cumbersome and error prone if there are multiple values to configure, or if this env var already
+contains configuration.
+
+## Implementation
+
+The implementation will add a helper to the tomcat buildpack that will look for launch variables named `BPL_TOMCAT_ENV_*`, e.g. `BPL_TOMCAT_ENV_POSTGRES_URL`, the helper 
+will take that variable, convert it to the system property `postgres.url` and append it to the environment variable `JAVA_TOOL_OPTIONS`.
+
+Pseudocode:
+```
+	systemProperty = envVar.removePrefix('BPL_TOMCAT_ENV_').toLowerCase().replaceAll('_', '.')
+```
+
+## Prior Art
+
+At present we've not found anything similar to this.
+
+## Unresolved Questions and Bikeshedding
+
+* Sensitive variables may be exposed in the application logs. We should ensure that `JAVA_TOOL_OPTIONS` is not logged, or is masked.
```

Comment:
I don't know if you can prevent this, the JVM logs the contents when it starts up. If we turn the values into system properties through a helper, that's going to add them to JAVA_TOOL_OPTIONS so they would show up in the logs.

- It might be possible to suppress that logging from the JVM.
- It might be possible to instead of adding system properties via JAVA_TOOL_OPTIONS, add them another way. Tomcat has `conf/catalina.properties` which is essentially a list of system properties that get set when Tomcat starts up. We might be able to have the helper write them to that file. We should be able to locate `$CATALINA_BASE/conf/catalina.properties` and append to it. The main trick would be that we'd need to make that file group writable at build time. Because the app can be run as a different user, but will always have the same group.

Reply:
I think that appending the system properties to `$CATALINA_BASE/conf/catalina.properties` is probably the best way to go, we can do that at startup via a helper (as you mentioned need to make sure that is group writable)